lib/ecto/changeset.ex

```diff
@@ -1569,6 +1569,61 @@ defmodule Ecto.Changeset do
     %{changeset | changes: changes, errors: errors, valid?: valid?}
   end
 
+  @doc """
+  Update the existing association entry or entries as a change in the changeset.
+
+  For association with cardinality one the passed function will receive a changeset.
+  For association with many entries the passed function will receive a list of changesets.
+
+  Returned values or lists of values need to match what is described in
+  `put_assoc/4`. If the given value is not any of values listed there,
+  it will raise. Commonly the returned value(s) are expected to be changesets.
+
+  Lists of entries can be reordered, e.g. for forms rendered based on the parent
+  changeset. But persisting associations does not retain order unless maintained
+  separately – e.g. using `prepare_changes/2`.
+  """
+  def update_assoc(%Changeset{} = changeset, name, default, fun) do
+    update_relation(:assoc, changeset, name, default, fun)
+  end
+
+  @doc """
+  Update the existing embed entry or entries as a change in the changeset.
+
+  For embeds with cardinality one the passed function will receive a changeset.
+  For embeds with many entries the passed function will receive a list of changesets.
+
+  Returned values or lists of values need to match what is described in
+  `put_assoc/4`. If the given value is not any of values listed there,
+  it will raise. Commonly the returned value(s) are expected to be changesets.
+
+  Lists of entries can be reordered.
+  """
+  def update_embed(%Changeset{} = changeset, name, default, fun) do
+    update_relation(:embed, changeset, name, default, fun)
+  end
+
+  defp update_relation(_tag, %{types: nil}, _name, _default, _fun) do
+    raise ArgumentError, "changeset does not have types information"
+  end
+
+  defp update_relation(tag, changeset, name, default, fun) when is_function(fun, 1) do
+    %{data: data, types: types, changes: changes, errors: errors, valid?: valid?} = changeset
+    relation = relation!(:put, tag, name, Map.get(types, name))
+    existing = get_change(changeset, name) || get_field(changeset, name, default)
```

Comment:
`Map.update/4` and `Keyword.update/4` don't pass the `default` into the update function. I see the use-case for this here, but is it adequate to have different semantics under the same name?

Reply:
Perhaps it makes sense to remove default and the user can either check for nil or an empty list?

Reply:
Yeah, that might make the most sense.

Reply:
Shouldn't `get_field` be enough as it looks on changes first and then falls back to the struct?

```suggestion
    existing = get_field(changeset, name)
```

Reply:
It isn't and that's exactly one of the footguns I want to get rid of by adding this. `get_field` applies changes for relations and doesn't return changesets if there are changes for those related records. 

Consider this:
```elixir
    base_changeset =
      Changeset.change(%Author{posts: [%Post{id: 1, title: "hello"}]})
      |> Changeset.cast(%{posts: [%{id: 1, title: "world"}]}, [])
      |> Changeset.cast_assoc(:posts)

    existing_posts = Changeset.get_field(base_changeset, :posts)
    changeset = Changeset.put_assoc(base_changeset, :posts, existing_posts)

    assert Map.has_key?(changeset.changes, :posts)
```

This test `fails` for actually two intertwined reasons. `get_field` returns a list of structs (so no way to know what is a change or the base data anymore) and `put_assoc` also treats `structs` as input as a special case of replacing existing data, not updating it as well. This breaks even before attempting to modify the `existing_posts` in between.